In [56]:
import pandas as pd
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [57]:
data=pd.read_csv('/content/i5dem.csv')

In [58]:
data

product;month_cons;inventory;rec_qte
0                            1% Milk;14.067;3.188;4.215
1                             1% Milk;14.067;1.691;5.03
2                              1% Milk;14.067;3.147;3.8
3                             1% Milk;14.067;3.71;3.732
4                            1% Milk;14.067;2.906;4.069
...                                                 ...
2700  Grilled Scallops with Lemon Butter;10.806;1.18...
2701  Grilled Scallops with Lemon Butter;10.806;1.18...
2702  Grilled Scallops with Lemon Butter;10.806;2.22...
2703  Grilled Scallops with Lemon Butter;10.806;3.94...
2704  Grilled Scallops with Lemon Butter;10.806;2.34...

[2705 rows x 1 columns]

In [59]:
data = data["product;month_cons;inventory;rec_qte"].str.split(";", expand=True)

# Rename the columns
data.columns = ["product", "month_cons","inventory","rec_qte"]
data.to_csv('/content/modified_data.csv', index=False)

In [60]:
data

product month_cons inventory rec_qte
0                                1% Milk     14.067     3.188   4.215
1                                1% Milk     14.067     1.691    5.03
2                                1% Milk     14.067     3.147     3.8
3                                1% Milk     14.067      3.71   3.732
4                                1% Milk     14.067     2.906   4.069
...                                  ...        ...       ...     ...
2700  Grilled Scallops with Lemon Butter     10.806     1.186   3.605
2701  Grilled Scallops with Lemon Butter     10.806     1.187   3.332
2702  Grilled Scallops with Lemon Butter     10.806     2.224   3.306
2703  Grilled Scallops with Lemon Butter     10.806     3.945   2.578
2704  Grilled Scallops with Lemon Butter     10.806     2.347   3.192

[2705 rows x 4 columns]

In [61]:
row_1106 = data.loc[439]

In [62]:
row_1106

product       Chicken Fajitas
month_cons             10.217
inventory               3.863
rec_qte                 2.586
Name: 439, dtype: object

In [63]:
res=data

In [64]:
data.to_csv('/content/mod_data.csv', index=False)

In [65]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
data["encoded_product"] = label_encoder.fit_transform(data["product"])

In [66]:
data

product month_cons inventory rec_qte  \
0                                1% Milk     14.067     3.188   4.215   
1                                1% Milk     14.067     1.691    5.03   
2                                1% Milk     14.067     3.147     3.8   
3                                1% Milk     14.067      3.71   3.732   
4                                1% Milk     14.067     2.906   4.069   
...                                  ...        ...       ...     ...   
2700  Grilled Scallops with Lemon Butter     10.806     1.186   3.605   
2701  Grilled Scallops with Lemon Butter     10.806     1.187   3.332   
2702  Grilled Scallops with Lemon Butter     10.806     2.224   3.306   
2703  Grilled Scallops with Lemon Butter     10.806     3.945   2.578   
2704  Grilled Scallops with Lemon Butter     10.806     2.347   3.192   

      encoded_product  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
2700              201  
2701              201  
2702              201  
2703              201  
2704              201  

[2705 rows x 5 columns]

In [67]:
uni_prod = data[["product", "encoded_product"]].drop_duplicates()

# Create a dictionary mapping product names to their encoded values
product_mapping = dict(zip(uni_prod["product"], uni_prod["encoded_product"]))

In [68]:
df=data

In [69]:
#df["inventory"] = pd.to_numeric(df["inventory"], errors="coerce")


In [70]:
#df["inventory"]=df["inventory"].astype(float)

In [71]:
df["month_cons"] = pd.to_numeric(df["month_cons"], errors="coerce")
df["month_cons"]=df["month_cons"].astype(float)
df["rec_qte"] = pd.to_numeric(df["rec_qte"], errors="coerce")
df["rec_qte"]=df["rec_qte"].astype(float)

In [72]:
df = df.dropna()

In [73]:
df

product  month_cons inventory  rec_qte  \
0                                1% Milk      14.067     3.188    4.215   
1                                1% Milk      14.067     1.691    5.030   
2                                1% Milk      14.067     3.147    3.800   
3                                1% Milk      14.067      3.71    3.732   
4                                1% Milk      14.067     2.906    4.069   
...                                  ...         ...       ...      ...   
2700  Grilled Scallops with Lemon Butter      10.806     1.186    3.605   
2701  Grilled Scallops with Lemon Butter      10.806     1.187    3.332   
2702  Grilled Scallops with Lemon Butter      10.806     2.224    3.306   
2703  Grilled Scallops with Lemon Butter      10.806     3.945    2.578   
2704  Grilled Scallops with Lemon Butter      10.806     2.347    3.192   

      encoded_product  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
2700              201  
2701              201  
2702              201  
2703              201  
2704              201  

[2705 rows x 5 columns]

In [74]:
df["month_cons"].astype(float)

0       14.067
1       14.067
2       14.067
3       14.067
4       14.067
         ...  
2700    10.806
2701    10.806
2702    10.806
2703    10.806
2704    10.806
Name: month_cons, Length: 2705, dtype: float64

In [75]:
df["inventory"].astype(float)
df["rec_qte"].astype(float)

0       4.215
1       5.030
2       3.800
3       3.732
4       4.069
        ...  
2700    3.605
2701    3.332
2702    3.306
2703    2.578
2704    3.192
Name: rec_qte, Length: 2705, dtype: float64

In [76]:
df = df.reset_index(drop=True)

In [77]:
data=df

In [78]:
data

product  month_cons inventory  rec_qte  \
0                                1% Milk      14.067     3.188    4.215   
1                                1% Milk      14.067     1.691    5.030   
2                                1% Milk      14.067     3.147    3.800   
3                                1% Milk      14.067      3.71    3.732   
4                                1% Milk      14.067     2.906    4.069   
...                                  ...         ...       ...      ...   
2700  Grilled Scallops with Lemon Butter      10.806     1.186    3.605   
2701  Grilled Scallops with Lemon Butter      10.806     1.187    3.332   
2702  Grilled Scallops with Lemon Butter      10.806     2.224    3.306   
2703  Grilled Scallops with Lemon Butter      10.806     3.945    2.578   
2704  Grilled Scallops with Lemon Butter      10.806     2.347    3.192   

      encoded_product  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
2700              201  
2701              201  
2702              201  
2703              201  
2704              201  

[2705 rows x 5 columns]

In [79]:
df.drop("product",axis=1,inplace=True)

In [80]:
Y=df["rec_qte"]

In [81]:
Y

0       4.215
1       5.030
2       3.800
3       3.732
4       4.069
        ...  
2700    3.605
2701    3.332
2702    3.306
2703    2.578
2704    3.192
Name: rec_qte, Length: 2705, dtype: float64

In [82]:
X=df.drop("rec_qte",axis=1)

In [83]:
X

month_cons inventory  encoded_product
0         14.067     3.188                0
1         14.067     1.691                0
2         14.067     3.147                0
3         14.067      3.71                0
4         14.067     2.906                0
...          ...       ...              ...
2700      10.806     1.186              201
2701      10.806     1.187              201
2702      10.806     2.224              201
2703      10.806     3.945              201
2704      10.806     2.347              201

[2705 rows x 3 columns]

In [84]:
import pickle
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Scale the features using StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
with open('scaler.pkl', 'wb') as f:
    pickle.dump(scaler, f)
# Build the regression model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(16, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(8, activation='relu'),

    tf.keras.layers.Dense(1)
])




In [85]:
X_test

month_cons inventory  encoded_product
1106      10.290     3.368              248
439       10.217     3.863               91
801       11.781     2.135              180
296       14.499     1.679               63
1718      13.525     2.502              385
...          ...       ...              ...
1190      14.638     3.287              266
2369      13.011     2.113              526
2543      14.642     2.175              112
1904      10.684     2.241              423
445       12.599     1.426               93

[541 rows x 3 columns]

In [86]:
# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
model.fit(X_train_scaled, y_train, epochs=100, batch_size=16, verbose=0)

# Get predictions on new data
y_pred = model.predict(X_test_scaled)

# Print the predictions
print(y_pred)

17/17 [==============================] - 0s 1ms/step
[[2.6735332]
 [2.443585 ]
 [3.70056  ]
 [4.8888817]
 [4.1919303]
 [5.03526  ]
 [4.6764073]
 [4.3478746]
 [2.1206517]
 [2.127667 ]
 [3.5362606]
 [1.9864107]
 [2.220556 ]
 [4.8189716]
 [4.096218 ]
 [2.830781 ]
 [4.390974 ]
 [3.3492575]
 [3.3701003]
 [2.7488334]
 [4.4266796]
 [2.5760422]
 [3.4634395]
 [2.9656315]
 [3.1626186]
 [3.9318075]
 [2.8639607]
 [4.359743 ]
 [2.7988799]
 [3.7212632]
 [4.1648803]
 [1.9445391]
 [3.3726456]
 [4.574282 ]
 [4.772997 ]
 [4.2513986]
 [3.1033583]
 [4.6596675]
 [4.734783 ]
 [2.770053 ]
 [1.8586576]
 [2.454115 ]
 [1.9888206]
 [4.7864165]
 [3.2735643]
 [4.213213 ]
 [3.8226414]
 [4.0003915]
 [2.3028893]
 [3.048139 ]
 [1.9524907]
 [2.5159876]
 [3.2586777]
 [4.3689246]
 [4.3549123]
 [4.2435713]
 [3.495049 ]
 [4.71939  ]
 [3.1375346]
 [4.3599863]
 [2.080711 ]
 [2.9233975]
 [3.1456556]
 [2.7369986]
 [5.0751624]
 [2.3628201]
 [4.5750875]
 [4.6529207]
 [3.8813562]
 [4.417041 ]
 [3.80193  ]
 [2.8113842]
 [2.2070727

In [87]:
print(y_test
      )

1106    2.533
439     2.586
801     3.401
296     4.792
1718    4.084
        ...  
1190    4.664
2369    4.672
2543    4.735
1904    3.474
445     4.664
Name: rec_qte, Length: 541, dtype: float64


In [88]:
print(y_pred)

[[2.6735332]
 [2.443585 ]
 [3.70056  ]
 [4.8888817]
 [4.1919303]
 [5.03526  ]
 [4.6764073]
 [4.3478746]
 [2.1206517]
 [2.127667 ]
 [3.5362606]
 [1.9864107]
 [2.220556 ]
 [4.8189716]
 [4.096218 ]
 [2.830781 ]
 [4.390974 ]
 [3.3492575]
 [3.3701003]
 [2.7488334]
 [4.4266796]
 [2.5760422]
 [3.4634395]
 [2.9656315]
 [3.1626186]
 [3.9318075]
 [2.8639607]
 [4.359743 ]
 [2.7988799]
 [3.7212632]
 [4.1648803]
 [1.9445391]
 [3.3726456]
 [4.574282 ]
 [4.772997 ]
 [4.2513986]
 [3.1033583]
 [4.6596675]
 [4.734783 ]
 [2.770053 ]
 [1.8586576]
 [2.454115 ]
 [1.9888206]
 [4.7864165]
 [3.2735643]
 [4.213213 ]
 [3.8226414]
 [4.0003915]
 [2.3028893]
 [3.048139 ]
 [1.9524907]
 [2.5159876]
 [3.2586777]
 [4.3689246]
 [4.3549123]
 [4.2435713]
 [3.495049 ]
 [4.71939  ]
 [3.1375346]
 [4.3599863]
 [2.080711 ]
 [2.9233975]
 [3.1456556]
 [2.7369986]
 [5.0751624]
 [2.3628201]
 [4.5750875]
 [4.6529207]
 [3.8813562]
 [4.417041 ]
 [3.80193  ]
 [2.8113842]
 [2.2070727]
 [3.7883172]
 [4.757515 ]
 [2.8448637]
 [2.6743972]

In [89]:
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


In [90]:
mse

0.07660574371216412

In [91]:
r2

0.9107071851776287

In [92]:
y_test

1106    2.533
439     2.586
801     3.401
296     4.792
1718    4.084
        ...  
1190    4.664
2369    4.672
2543    4.735
1904    3.474
445     4.664
Name: rec_qte, Length: 541, dtype: float64

In [93]:
ww=y_test.values

In [94]:
ww

array([2.533, 2.586, 3.401, 4.792, 4.084, 5.467, 4.817, 4.362, 2.001,
       1.876, 3.222, 2.204, 2.052, 4.646, 4.44 , 3.077, 4.717, 2.955,
       3.805, 2.872, 5.072, 2.595, 3.122, 3.062, 3.498, 4.364, 3.057,
       3.925, 2.745, 3.614, 3.875, 1.98 , 3.66 , 5.003, 4.809, 3.858,
       3.197, 4.172, 4.43 , 3.034, 1.65 , 2.801, 1.77 , 5.174, 3.487,
       4.485, 4.289, 4.154, 2.169, 3.432, 2.016, 2.378, 3.191, 3.972,
       3.853, 3.812, 3.559, 4.612, 3.457, 3.939, 1.809, 3.126, 2.916,
       3.105, 4.67 , 2.487, 4.783, 4.139, 4.08 , 4.787, 3.575, 2.693,
       2.138, 3.534, 5.199, 2.654, 2.513, 4.257, 1.877, 2.565, 2.826,
       1.738, 3.076, 4.194, 2.75 , 1.996, 3.063, 4.126, 3.764, 4.557,
       4.878, 3.071, 4.657, 3.968, 2.302, 2.981, 2.495, 4.294, 4.218,
       3.48 , 3.569, 3.82 , 3.313, 4.383, 4.776, 2.003, 2.39 , 3.628,
       2.769, 1.836, 3.911, 5.459, 4.171, 3.916, 3.838, 2.791, 2.235,
       2.927, 2.118, 4.363, 4.179, 1.952, 3.102, 3.274, 4.574, 4.092,
       1.633, 4.083,

In [95]:
ww[0]

2.533

In [96]:
l=[]
for i in range(len(ww)):
  l.append(abs(ww[i]-y_pred[i][0]))
print(max(l))

0.7432393875122072


In [97]:
print(ww[4])
print(y_pred[4][0])

4.084
4.1919303


In [98]:
X_test

month_cons inventory  encoded_product
1106      10.290     3.368              248
439       10.217     3.863               91
801       11.781     2.135              180
296       14.499     1.679               63
1718      13.525     2.502              385
...          ...       ...              ...
1190      14.638     3.287              266
2369      13.011     2.113              526
2543      14.642     2.175              112
1904      10.684     2.241              423
445       12.599     1.426               93

[541 rows x 3 columns]

In [99]:
data

month_cons inventory  rec_qte  encoded_product
0         14.067     3.188    4.215                0
1         14.067     1.691    5.030                0
2         14.067     3.147    3.800                0
3         14.067      3.71    3.732                0
4         14.067     2.906    4.069                0
...          ...       ...      ...              ...
2700      10.806     1.186    3.605              201
2701      10.806     1.187    3.332              201
2702      10.806     2.224    3.306              201
2703      10.806     3.945    2.578              201
2704      10.806     2.347    3.192              201

[2705 rows x 4 columns]

In [100]:
data[data['encoded_product']==248]

month_cons inventory  rec_qte  encoded_product
1105       10.29     2.352    2.998              248
1106       10.29     3.368    2.533              248
1107       10.29     3.459    2.716              248
1108       10.29     1.366    3.105              248
1109       10.29     2.137    3.113              248

In [101]:
row_1106 = data.loc[801]

In [102]:
row_1106

month_cons         11.781
inventory           2.135
rec_qte             3.401
encoded_product       180
Name: 801, dtype: object

product       Chicken Fajitas

month_cons             10.217
inventory               3.863
rec_qte                 2.586

In [103]:
encoded_product = res.loc[res['product'] =="Light Sour Cream", 'encoded_product'].values[0]

In [104]:
encoded_product

248

In [108]:
from keras import models
import numpy as np

def get_predictions( month_cons, inventory, product,model=model, scaler=scaler):

    encoded_product = res.loc[res['product'] == product, 'encoded_product'].values[0]
    input_data = np.array([[month_cons, inventory, encoded_product]])


    scaled_input_data = scaler.transform(input_data)


    predictions = model.predict(scaled_input_data)
    return predictions

prediction = get_predictions(10.217,3.863,"Chicken Fajitas")
print("Predicted rec_qte:", prediction[0][0])


1/1 [==============================] - 0s 19ms/step
Predicted rec_qte: 2.443585


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [ ]:
y_test

In [ ]:
import pickle

In [ ]:
pickle.dump(model,open('_model.pkl','wb'))
pickle.dump(res,open('dta.pkl','wb'))

In [ ]:
pickle.dump(scaler,open('scaler.pkl','wb'))